### PR TITLE
add impersonate user config for ldap authentication

### PR DESCRIPTION
**Pull request by aezomz**
## Overview
Add impersonate_user config to ldap credential authentication mode. Issue created #163 
Sometime we need to authenticate with service account but impersonate the user running the service instead.

Please let me know if there is a better way for this. 
I am not able to run the test on my local though :( 

## Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change
- [x] I have run `changie new` to [create a changelog entry](https://github.com/starburstdata/dbt-trino/blob/master/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/adapters/trino/connections.py [hunk 1]
```diff
@@ -408,7 +409,7 @@ def open(cls, connection):
         trino_conn = trino.dbapi.connect(
             host=credentials.host,
             port=credentials.port,
-            user=credentials.user,
+            user=credentials.impersonate_user or credentials.user,
```
Comment (findinpath) [inline dbt/adapters/trino/connections.py:412]: pls add a corresponding test in `test_adapter.py`

File: README.md [hunk 2]
```diff
@@ -44,6 +44,7 @@ A dbt profile can be configured to run against Trino using the following configu
 | method                         | The Trino authentication method to use                                                                       | Optional (default is `none`, supported methods are `ldap`, `kerberos`, `jwt`, `oauth` or `certificate`) | `none` or `kerberos`             |
 | user                           | Username for authentication                                                                                  | Optional (required if `method` is `none`, `ldap` or `kerberos`)                                         | `commander`                      |
 | password                       | Password for authentication                                                                                  | Optional (required if `method` is `ldap`)                                                               | `none` or `abc123`               |
+| impersonate_user               | impersonate user, as X-Trino-User                                                                            | Optional (applicable if `ldap`)                                                                         | `impersonated_tom`               |
```
Comment (hovaesco) [inline README.md:47]: Description: Username override, used for impersonation.

Comment (hovaesco) [reply README.md:47]: Parameter name: `impersonation_user`

Comment (aezomz) [reply README.md]: Updated :)

File: tests/unit/test_adapter.py [hunk 3]
```diff
@@ -238,6 +238,7 @@ def test_ldap_authentication(self):
                 "method": "ldap",
                 "schema": "dbt_test_schema",
                 "user": "trino_user",
+                "impersonation_user": "impersonated_user",
```
Comment (aezomz) [inline tests/unit/test_adapter.py:241]: Is there a place I can assert the user we pass into dbapi ?

File: dbt/adapters/trino/connections.py (hunk 1 above)
Comment (aezomz) [reply dbt/adapters/trino/connections.py]: Sure, i am not sure where to assert the dbapi user. Can you point me to it?

File: tests/unit/test_adapter.py (hunk 3 above)
Comment (hovaesco) [reply tests/unit/test_adapter.py:241]: Just please add `assertEqual` for `impersonation_user` below.

**Comment by aezomz** [comment]
Squashed. thanks!

**Comment by findinpath** [comment]
@aezomz 

Commit comment:

`added impersonate user for ldap auth` -> `Add ability to impersonate user for LDAP authentication`

Do add an alternative test case for the impersonation so that we have for ldap:
- a test without impersonation
- a test with impersonation

**Comment by aezomz** [comment]
Is somethign like this ok? @findinpath 
```
    def test_ldap_authentication(self):
        connection = self.acquire_connetion_with_profile(
            {
                "type": "trino",
                "catalog": "trinodb",
                "host": "database",
                "port": 5439,
                "method": "ldap",
                "schema": "dbt_test_schema",
                "user": "trino_user",
                "password": "trino_password",
                "cert": "/path/to/cert",
                "http_headers": {"X-Trino-Client-Info": "dbt-trino"},
                "session_properties": {
                    "query_max_run_time": "5d",
                    "exchange_compression": True,
                },
            }
        )
        credentials = connection.credentials
        self.assertIsInstance(credentials, TrinoLdapCredentials)
        self.assert_default_connection_credentials(credentials)
        self.assertEqual(credentials.http_scheme, HttpScheme.HTTPS)
        self.assertEqual(credentials.cert, "/path/to/cert")
        self.assertEqual(credentials.impersonation_user, None)

    def test_ldap_authentication_impersonation(self):
        connection = self.acquire_connetion_with_profile(
            {
                "type": "trino",
                "catalog": "trinodb",
                "host": "database",
                "port": 5439,
                "method": "ldap",
                "schema": "dbt_test_schema",
                "user": "trino_user",
                "impersonation_user": "impersonated_user",
                "password": "trino_password",
                "cert": "/path/to/cert",
                "http_headers": {"X-Trino-Client-Info": "dbt-trino"},
                "session_properties": {
                    "query_max_run_time": "5d",
                    "exchange_compression": True,
                },
            }
        )
        credentials = connection.credentials
        self.assertIsInstance(credentials, TrinoLdapCredentials)
        self.assert_default_connection_credentials(credentials)
        self.assertEqual(credentials.http_scheme, HttpScheme.HTTPS)
        self.assertEqual(credentials.cert, "/path/to/cert")
        self.assertEqual(credentials.impersonation_user, "impersonated_user")
```

**Comment by findinpath** [comment]
@aezomz  please avoid code duplication .

Note that using `connection.handle` will open the connection (see `dbt.adapters.trino.connections.TrinoConnectionManager.open`)


```
def test_ldap_authentication_impersonation(self):
        connection = self.acquire_connetion_with_profile(
            {
                "type": "trino",
                "catalog": "trinodb",
                "host": "database",
                "port": 5439,
                "method": "ldap",
                "schema": "dbt_test_schema",
                "user": "trino_user",
                "impersonation_user": "impersonated_user",
                "password": "trino_password",
                "cert": "/path/to/cert",
                "http_headers": {"X-Trino-Client-Info": "dbt-trino"},
                "session_properties": {
                    "query_max_run_time": "5d",
                    "exchange_compression": True,
                },
            }

            connection.handle
```


Afterwards you can see exactly what user is being used in the connection

```
connection.handle.handle.user
```

**Review by mdesmet — APPROVED** [review]

**Pull request merged**